### PR TITLE
Added missing error check that could cause kubelet to crash

### LINKED ISSUE #60861: FlexVolume probe race condition potentially crashes kubelet
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
If a new file is introduced in the FlexVolume driver directory then quickly deleted, a race condition would occur causing segmentation violation and crashing kubelet.

**What you expected to happen**:
No crashes.

**How to reproduce it (as minimally and precisely as possible)**:
It is a race condition. A file needs to be created then deleted many times until the kubelet crashes.

**Anything else we need to know?**:
Patch:
```patch
--- a/pkg/volume/flexvolume/probe.go
+++ b/pkg/volume/flexvolume/probe.go
@@ -176,9 +176,11 @@ func (prober *flexVolumeProber) updateProbeNeeded() {
 // on its parent directory.
 func (prober *flexVolumeProber) addWatchRecursive(filename string) error {
        addWatch := func(path string, info os.FileInfo, err error) error {
-               if info.IsDir() {
-                       if err := prober.watcher.AddWatch(path); err != nil {
-                               glog.Errorf("Error recursively adding watch: %v", err)
+               if err != nil {
+                       if info.IsDir() {
+                               if err := prober.watcher.AddWatch(path); err != nil {
+                                       glog.Errorf("Error recursively adding watch: %v", err)
+                               }
                        }
                }
                return nil
```

crash log:
```
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: panic: runtime error: invalid memory address or nil pointer dereference
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x310e75f]
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: goroutine 113 [running]:
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/volume/flexvolume.(*flexVolumeProber).addWatchRecursive.func1(0xc422965000, 0x3d, 0x0, 0x0, 0x9afbf60, 0xc421d04030, 0x0, 0x10)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:179 +0x2f
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: path/filepath.Walk(0xc422965000, 0x3d, 0xc423ce80e0, 0xc420a515d0, 0xc420c74270)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /usr/local/go/src/path/filepath/path.go:396 +0x8f
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/util/filesystem.DefaultFs.Walk(0xc422965000, 0x3d, 0xc423ce80e0, 0x41fe020, 0x1)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/filesystem/defaultfs.go:86 +0x3f
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/util/filesystem.(*DefaultFs).Walk(0x9e89e28, 0xc422965000, 0x3d, 0xc423ce80e0, 0x1d, 0x0)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         <autogenerated>:27 +0x63
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/volume/flexvolume.(*flexVolumeProber).addWatchRecursive(0xc42062dab0, 0xc422965000, 0x3d, 0x1d, 0x0)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:186 +0x89
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/volume/flexvolume.(*flexVolumeProber).handleWatchEvent(0xc42062dab0, 0xc422965000, 0x3d, 0xc400000001, 0x43d055, 0xc42138dee8)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:146 +0x14d
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/volume/flexvolume.(*flexVolumeProber).initWatcher.func1(0xc422965000, 0x3d, 0xc400000001)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/flexvolume/probe.go:193 +0x4a
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: k8s.io/kubernetes/pkg/util/filesystem.(*fsnotifyWatcher).Run.func1(0xc420b003a0)
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/filesystem/watcher.go:80 +0x145
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]: created by k8s.io/kubernetes/pkg/util/filesystem.(*fsnotifyWatcher).Run
Mar 05 23:38:22 k8s-agentpool0-31135900-0 docker[7361]:         /workspace/anago-v1.8.7-beta.0.34+b30876a5539f09/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/filesystem/watcher.go:88 +0x3f
Mar 05 23:38:23 k8s-agentpool0-31135900-0 systemd[1]: kubelet.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Mar 05 23:38:23 k8s-agentpool0-31135900-0 systemd[1]: kubelet.service: Unit entered failed state.
Mar 05 23:38:23 k8s-agentpool0-31135900-0 systemd[1]: kubelet.service: Failed with result 'exit-code'.
```

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.7 - applies to all
- Cloud provider or hardware configuration: Azure ACS
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.3 LTS`
- Kernel (e.g. `uname -a`): `Linux k8s-agentpool0-31135900-0 4.13.0-1011-azure #14-Ubuntu SMP Thu Feb 15 16:15:39 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by technicianted**
**What this PR does / why we need it**:
Adds missing error check. An error can happen due to a race condition when watched files change, or become inaccessible. This can happen if a file was added to the driver directory then quickly removed, in which case the callback will be called with non-nil `err` and nil `info`, which is not checked, causing kubelet to crash.

**Which issue(s) this PR fixes**:
Fixes #60861

**Special notes for your reviewer**:

**Release note**:

```release-note
Fixed missing error checking that could cause kubelet to crash in a race condition.
```

### CODE REVIEW

**Comment by dims** [comment]
/milestone v1.10

(per @saad-ali 's comment in issue)

**Comment by dims** [comment]
@technicianted hmm, you have signed the CLA already?

**Comment by BenTheElder** [comment]
Hi, GitHub doesn't seem to associate the commit with your account, have you verified the email address used in the commit with GitHub?

**Comment by technicianted** [comment]
@BenTheElder seems like it. Not sure how to fix the commit with the correct email.

**Comment by BenTheElder** [comment]
FWIW GitHub has an article here:
https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/
I checked out your PR and the commit seems to have an intel.com email which I'd guess is correct, but your email isn't public so I can't be sure.

**Comment by dims** [comment]
@technicianted it's easy to switch the email address for example see https://gist.github.com/trey/9588090

File: pkg/volume/flexvolume/probe.go [hunk 1]
```diff
@@ -176,9 +176,11 @@ func (prober *flexVolumeProber) updateProbeNeeded() {
 // on its parent directory.
 func (prober *flexVolumeProber) addWatchRecursive(filename string) error {
 	addWatch := func(path string, info os.FileInfo, err error) error {
-		if info.IsDir() {
-			if err := prober.watcher.AddWatch(path); err != nil {
-				glog.Errorf("Error recursively adding watch: %v", err)
+		if err == nil {
+			if info.IsDir() {
```
Comment (verult) [inline pkg/volume/flexvolume/probe.go]: What about `err == nil && info.IsDir()`? Reduces nesting

**Comment by technicianted** [comment]
@dims I signed the CLA. I made sure my commits have the correct GitHub verified email.

**Comment by dims** [comment]
@technicianted please reach out to the helpdesk - https://github.com/kubernetes/kubernetes/issues/27796#issuecomment-356006074

here's what i see in the PR:
```
commit 0fb8072b1956cec112f9b961e324a21d69131d39 (HEAD)
Author: technicianted <osama.sorour@eformations.net>
Date:   Fri Mar 9 21:03:22 2018 -0800

    reduce nesting

commit 659d9df117531069026a2c31310702d6dee5d9f0
Author: technicianted <osama.sorour@eformations.net>
Date:   Thu Mar 8 21:39:22 2018 -0800

    added missing error check

```

**Comment by saad-ali** [comment]
Once this is merged, please make sure to cherry pick it back to relevant prior release branches (1.9, etc.)

**Comment by saad-ali** [comment]
@technicianted please remember to cherry pick this PR back to relevant branches (1.9, etc.)

**Comment by technicianted** [comment]
@saad-ali please help with cherrypicking

**Comment by dims** [comment]
@technicianted can you please add a user facing release note in the body of the issue under "Release note:" section above? (currently has NONE there). the branch managers will ask for details.

**Comment by technicianted** [comment]
@dims added

**Comment by dims** [comment]
@technicianted thanks. i lined up the cherry picks for you.

**Pull request merged**